### PR TITLE
Add DiffOpFactory

### PR DESCRIPTION
**Pull request by addshore**
This is a CP from WikibaseLib
(with a minor alteration)

### CODE REVIEW

File: src/Diff/DiffOpFactory.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace Wikibase;
+
+use Diff\DiffOp\DiffOp;
+use Diff\DiffOpFactory as DiffDiffOpFactory;
```
Comment (JeroenDeDauw) [inline src/Diff/DiffOpFactory.php]: :)

File: src/Diff/DiffOpFactory.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace Wikibase;
```
Comment (JeroenDeDauw) [inline src/Diff/DiffOpFactory.php]: NS. The code is broken as it, since it won't get autoloaded. Lack of test is bad...

**Comment by JeroenDeDauw** [comment]
What is the minor alteration?

File: src/Diff/DiffOpFactory.php (hunk 1 above)
Comment (addshore) [inline src/Diff/DiffOpFactory.php]: This was the minor alteration ;) So that this class doesn't have to be called WikibaseDiffOpFactory

**Comment by addshore** [comment]
See the inline comment, I renamed the class basically ;)

**Comment by JeroenDeDauw** [comment]
As far as I can tell this test was copied from elsewhere and just tests the behaviour of the superclass (which is concrete and already tested). I'd rather see a test for the behaviour the class itself adds.

File: src/Diff/DiffOpFactory.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+namespace Wikibase\DataModel\Services\Diff;
+
+use Diff\DiffOp\DiffOp;
+use Diff\DiffOpFactory as DiffDiffOpFactory;
+use InvalidArgumentException;
+
+/**
+ * Class for changes that can be represented as a Diff.
+ *
+ * @since 1.1
```
Comment (thiemowmde) [inline src/Diff/DiffOpFactory.php]: 1.2?

Comment (addshore) [inline src/Diff/DiffOpFactory.php]: Yep, this would now be going in for 1.2

Comment (JeroenDeDauw) [inline src/Diff/DiffOpFactory.php]: Updated

**Comment by thiemowmde** [comment]
-1 for having two classes with the exact same name.

Not sure what's wrong with the test. Without this the class would have zero coverage.

**Comment by JeroenDeDauw** [comment]
We are not adding a new class or renaming one. This is a move from Wikibase Lib to this component. They already have the same name.

My concern about the test has already been addressed

**Comment by thiemowmde** [comment]
Sure this class got renamed, @addshore talked about this above. It was called `WikibaseDiffOpFactory` in Lib. It's basically `class \Wikibase\DataModel\Services\Diff\DiffOpFactory extends \Diff\DiffOpFactory` now and I disagree with two classes, one extending the other, having the exact same name. The extended one should make it's purpose clear with a more specific name.

**Comment by JeroenDeDauw** [comment]
#fail >_> addshore even wrote that in the comment above.

**Comment by addshore** [comment]
The purpose is clear through the namespace of the class.
I can't see any case where these 2 classes are going to be used side by side.
The Wikibase prefix to me seems quite pointless, but if others want it I'll add it back

**Comment by thiemowmde** [comment]
~~I just realized that this class being a subclass of `\Diff\DiffOpFactory` is completely pointless. It does not use **any** code from it's parent except for the trivial assertion. Should the parent class be abstract (and renamed accordingly)?~~

~~Even simpler, just ditch the `extends` and ditch the whole base class. It's completely unused in our code base. Then you are free to name the remaining class after the former base class.~~

**Comment by thiemowmde** [comment]
Even simpler, just ditch the `extends` and ditch the whole base class. It's completely unused in out code base. Then you are free to name the remaining class after the former base class.

**Comment by JeroenDeDauw** [comment]
I guess you did not see the `return parent::newFromArray( $diffOp );`?

Removing the baselcass from Diff is somewhat odd as it mirrors the `toArray` behaviour of the operations defined in Diff.

**Comment by thiemowmde** [comment]
Yes, I missed the `parent` call. Ok, then this class should really have a more specific name that reflects the purpose and difference to the base class.

**Comment by addshore** [comment]
Well, if anyone has any suggestion I will happily amend this....
My thoughts on the name can already be found @ https://github.com/wmde/WikibaseDataModelServices/pull/22#issuecomment-130344511

**Comment by JeroenDeDauw** [comment]
I'm fine with this as it is if these things will not be used together, which indeed seems likely. Or with renaming it. +2 from my side, feel free to merge. A follow up can be submitted if someone wants to change it.

**Comment by thiemowmde** [comment]
+2 if #45 is merged first.

**Comment by addshore** [comment]
BAH, needs a rebase with that merged ontop apparently....

**Comment by thiemowmde** [comment]
Rebased. +2 from my side now that #45 is merged in here, so feel free to merge.

**Pull request merged**